README.md
# Hazel - Linux Chatbot and System Assistant
<hr>

### What is it about?
Hazel aims to be a beginner-friendly, assistive chatbot. Many of us know the benefits of using a free OS (like Linux, BSD, etc), yet we never use it as our primary OS. It's a general knowledge now that proprietary software and vensors force us to use their technology without giving us a choice. In my personal experience, more than cost-effectiveness or free content I felt choices very necessary and appealing. <br/>Hazel enable you enjoy this freedom without having to bother about the painful learning curve associated with a linux system. Hazel simply lets you enjoy the "view" while it runs "linux commands in the background". Users need not bother about having to learn how it works or how to do things. Hazel automates more than what an average user would want out of their linux machine.

### Getting started!!!
There are no complexity associated with using this software. The only skill you'll ever need to use your linux system is the ability to read and type basic english. You can ask hazel to do some actions for you and be done with it. You can perform almost all the basic tasks you want by chating with this bot.<br>If you are *an absolute beginner *or* is interested in learning Linux* then join now. Grab **Manjaro Linux** to get started or any other linux to test features other than package management.

### Really getting started...
You might need to grab the following to start using Hazel<br>
* **Manjaro Linux** - I started this project exclusively for Arch-based linux since I found it more beginner friendly and appealing than other linux distributions out there (Sorry Ubuntu and other pop distros! That's a naked truth)

* **Python2 and pip** - Ummh!!! Not for coding or any other geek suff. Install these to make sure the testing release of Hazel works properly as long as a stable version is not released. 
 
 > sudo pacman -S python2-pip  
* **Ability to read and type in English** and the ability to open  **terminal** all by yourself.... LOL! JK...


Download [Manjaro Linux](http://manjaro.org/get-manjaro/)
### Install and Run Hazel on your system...

If you are a new user, download git first<br>
> sudo pacman -S git <br>

Clone the repository to your deivce home :
> git clone https://github.com/illuminati-RV/hazel-linux_chatbot_and_assistant.git <br>

Change present location to the project folder<br>
> cd hazel-linux_chatbot_and_assistant/hazel/ <br>

Setup nltk_data and other requirements
>./setup.sh<br>

[Optional] This script only installs essential nltk data, to get the whole nltk data, open a python shell<br>
> python2<br>
> > import nltk<br>
> > nltk.download()

Run the main.py file to get started with Hazel
> python2 main.py


<br><br>
### About this project
This is a small project initiated to bring more people to the FOSS world. You can modify the code to your liking and preferences for private usage. If you wish to publish your work to the public, make sure to include the license and source code along with the publication.<br>If you are interested in learning python or coding in python, get started with this project. You can contact me for more details.<br><center> **CODE . SHARE . PROSPER** </center>


hazel/history.txt

# 2018-02-20 20:12:55.229802
+hello there

# 2018-02-20 20:13:54.524485
+hi

# 2018-02-20 20:13:59.091426
+how are you

# 2018-02-20 20:14:13.083302
+i want to download firefox

# 2018-02-20 20:14:34.553148
+q

# 2018-02-20 20:31:08.504236
+hi

# 2018-02-20 20:31:48.564802
+q

# 2018-02-20 20:31:53.971130
+hello

# 2018-02-20 20:32:01.371187
+q

# 2018-02-20 20:33:16.618066
+hello

# 2018-02-20 20:33:18.450715
+hi

# 2018-02-20 20:35:28.504487
+what is my ip address

# 2018-02-20 20:36:59.577853
+q

# 2018-02-20 20:38:29.109354
+hi

# 2018-02-20 20:38:37.420806
+how are you

# 2018-02-20 20:38:44.204910
+install firefox

# 2018-02-20 20:39:01.044878
+q

# 2018-02-20 20:49:07.533292
+time now

# 2018-02-20 20:49:19.090843
+update system

# 2018-02-20 20:50:06.256942
+who is

# 2018-02-20 20:50:30.508437
+who is gandhi

# 2018-02-20 20:52:57.966711
+q

# 2018-02-20 20:53:14.000471
+i want to install firefox

# 2018-02-20 20:53:36.720216
+get my ip address

# 2018-02-20 20:53:43.616356
+q

# 2018-02-20 20:54:39.939855
+my ip address

# 2018-02-20 20:54:45.543629
+q

# 2018-02-20 20:55:12.544337
+install firefox

# 2018-02-20 20:55:40.121877
+remove firefox

# 2018-02-20 20:55:54.818116
+install chromium

# 2018-02-20 20:56:19.520762
+i want to install chrome

# 2018-02-20 20:56:36.843402
+install chromium

# 2018-02-20 20:56:49.548653
+i want to install chromium

# 2018-02-20 20:57:35.018414
+quit

# 2018-02-20 20:57:39.455969
+i want to install chromium

# 2018-02-20 20:58:19.635923
+remoe chromium

# 2018-02-20 20:58:29.558870
+remove chromium

# 2018-02-20 20:58:52.062916
+quit

# 2018-02-20 20:59:40.950470
+upodate systen

# 2018-02-20 20:59:50.498496
+update system

# 2018-02-20 23:33:39.659217
+update my system

# 2018-02-20 23:55:06.869832
+how much ram is available

# 2018-02-20 23:57:50.290319
+q

# 2018-02-21 00:00:24.752285
+storage space left

# 2018-02-21 00:03:17.332220
+q

# 2018-02-21 00:06:56.670786
+data usage

# 2018-02-21 00:56:05.065495
+what is my username

# 2018-02-21 00:56:32.538748
+q

# 2018-02-21 00:56:37.260439
+what is my username

# 2018-02-21 00:56:46.405528
+list all users

# 2018-02-21 06:36:21.486817
+hi

# 2018-02-21 06:36:28.252693
+what is my username

# 2018-02-21 06:36:35.704400
+list all users

# 2018-02-21 06:36:41.157600
+q

# 2018-02-21 06:38:48.992049
+list all users

# 2018-02-21 06:39:16.945626
+q

# 2018-02-21 06:39:20.414157
+list all users

# 2018-02-21 06:39:29.918230
+q

# 2018-02-21 06:40:11.477108
+list all users

# 2018-02-21 06:40:18.856219
+q

# 2018-02-21 06:46:24.589309
+what is my ram usage

# 2018-02-21 06:47:10.003843
+q

# 2018-02-21 06:47:26.968073
+eam usage

# 2018-02-21 06:47:32.464599
+ram usage

# 2018-02-21 06:47:49.249179
+q

# 2018-02-21 06:47:52.489825
+ram usage

# 2018-02-21 06:49:35.212874
+q

# 2018-02-21 06:49:40.731889
+ram usage

# 2018-02-21 06:49:59.809301
+q

# 2018-02-21 06:50:05.819947
+ram usage

# 2018-02-21 06:50:37.486748
+quit

# 2018-02-21 06:50:42.522606
+ram usage

# 2018-02-21 06:52:05.642210
+q

# 2018-02-21 06:52:19.485337
+how much ram is available

# 2018-02-21 06:52:25.857953
+q

# 2018-02-21 09:40:50.338146
+hi

# 2018-02-21 12:44:21.423663
+what is my battery status

# 2018-02-21 12:52:28.061666
+how much ram is used

# 2018-02-21 13:22:06.684272
+change date

# 2018-02-21 13:23:07.604568
+q

# 2018-02-21 13:23:13.841073
+change date

# 2018-02-21 13:36:25.075845
+q

# 2018-02-21 13:36:44.020017
+change date

# 2018-03-03 00:00:11.037571
+q

# 2018-02-21 14:25:27.763559
+internet usafe

# 2018-02-21 14:25:32.872777
+internet usage

# 2018-02-21 14:27:32.484353
+q

# 2018-02-21 14:27:45.901468
+internet usage

# 2018-02-21 14:28:01.260603
+battery status

# 2018-02-21 19:23:01.663727
+q

# 2018-02-21 19:23:58.651424
+quit

# 2018-02-21 19:24:53.457249
+hello hazel

# 2018-02-21 19:25:01.168501
+how are you today

# 2018-02-21 19:25:05.661669
+am fine

# 2018-02-21 19:25:08.222801
+q

# 2018-02-21 19:25:58.305094
+quit

# 2018-02-21 19:26:13.623540
+get firefox

# 2018-02-21 19:26:19.896870
+install firefox

# 2018-02-21 19:27:56.091253
+remove firefox

# 2018-02-21 19:28:00.809719
+q

# 2018-02-21 19:23:01.663727
+q

# 2018-02-21 19:23:58.651424
+quit

# 2018-02-21 19:24:53.457249
+hello hazel

# 2018-02-21 19:25:01.168501
+how are you today

# 2018-02-21 19:25:05.661669
+am fine

# 2018-02-21 19:25:08.222801
+q

# 2018-02-21 19:25:58.305094
+quit

# 2018-02-21 19:26:13.623540
+get firefox

# 2018-02-21 19:26:19.896870
+install firefox

# 2018-02-21 19:27:56.091253
+remove firefox

# 2018-02-21 19:28:00.809719
+q

# 2018-02-21 19:37:31.614618
+i am in temp directory

# 2018-02-21 19:37:40.318216
+q

# 2018-02-21 19:39:09.546094
+1

# 2018-02-21 19:39:12.049802
+4

# 2018-02-21 19:39:16.262390
+q

# 2018-02-21 19:39:43.021554
+install firefox

# 2018-02-21 19:40:00.971015
+q


hazel/main.py
#!/usr/bin/python

##########################################################################################################################

### Importing important python packages required by Hazel ###

from __future__ import print_function
import os
import sys
import time
import socket
import signal
import psutil
import random
import getpass
import difflib
import readline
import operator
import platform
import datetime
from bs4 import *
from nltk import *
import wikipedia as w
from gmail import *
from wikisearch import *
from systeminfo import *
from wikiapi import WikiApi
from hazel_chatter import *
from package_surfer import *
from textblob import TextBlob
from nltk.corpus import stopwords
from prompt_toolkit import prompt
from collections import *
#from collections import OrderedDict
from difflib import SequenceMatcher
from nltk.tokenize import word_tokenize
from prompt_toolkit.history import FileHistory
from vocabulary.vocabulary import Vocabulary as vb
from prompt_toolkit.contrib.completers import WordCompleter
from prompt_toolkit.auto_suggest import AutoSuggestFromHistory

##########################################################################################################################

### FUNCTION TO HANDLE EVENTS SUCH AS ctrl+C OR ctrl+Z TO PREVENT THE PROGRAM FROM TERMINATION ###

def sigint_handler(signum, frame):
    print ('') # Display message on occurance of below events

signal.signal(signal.SIGINT, sigint_handler) # handle Ctrl+C on event
signal.signal(signal.SIGTSTP, sigint_handler) # handle Ctrl+Z on event

def data(i):
    return i

global callme
callme = "there"


##########################################################################################################################


### FUNCTION TO IMPLEMENT FUNCTIONS  OTHER THAN PACKAGE MANAGEMENT, REDIRECTED FROM MAIN() FUNCTION BELOW ###
### UNFORMATTED AND NEEDS A MAJOR CHANGE TO WORK AS EXPECTED, CURRENTLY IN STAGE 1 ####

def others(tokens, message, unstopped_tokens, chat): # Funtion defention of Others, passed arguements are the tokenized message and original message or user input

    global callme

    ### WIKIPEDIA SEARCH ####
    
    if "tell" in unstopped_tokens and "about" in unstopped_tokens:
        wiki(unstopped_tokens, message)

    elif "tell" in unstopped_tokens and "more" in unstopped_tokens:
        wiki(unstopped_tokens, message)

    elif "change" in tokens and "date" in tokens:
        set_date()

    elif "date" in tokens:
        datenow(tokens)

    elif "time" in tokens:
        timenow(tokens)

    elif "my" in unstopped_tokens and "username" in unstopped_tokens or "my" in unstopped_tokens and "user" in unstopped_tokens and "name" in unstopped_tokens:
        if "change" in tokens:
            socli(chat, unstopped_tokens)
        username = getpass.getuser()
        print("\nHazel : Your username on this system is ",username)

    elif "look" in unstopped_tokens and "for" in unstopped_tokens:
        wiki(unstopped_tokens, message)

    elif "battery" in tokens and "status" in tokens or "battery" in tokens and "charge" in tokens or "battery" in tokens and "status" in tokens:
        battery = psutil.sensors_battery()
        plugged = battery.power_plugged
        percent = str(battery.percent)
        if plugged==False: plugged="Not Plugged In"
        else: plugged="Plugged In"
        print("\nHazel : Your battery percentage is ", percent + "% | " + plugged)

    elif "ram" in unstopped_tokens and "usage" in unstopped_tokens or "ram" in tokens and "available" in tokens or "ram" in unstopped_tokens and "used" in unstopped_tokens:
        print("\033[1;34;1m")
        mem=str(os.popen('free -t -m').readlines())
        T_ind=mem.index('T')
        mem_G=mem[T_ind+14:-4]
        S1_ind=mem_G.index(' ')
        mem_T=mem_G[0:S1_ind]
        mem_G1=mem_G[S1_ind+8:]
        S2_ind=mem_G1.index(' ')
        mem_U=mem_G1[0:S2_ind]
        mem_F=mem_G1[S2_ind+8:]
        print('Total Memory = ' + mem_T + ' MB')    
        print('Free Memory = ' + mem_F + ' MB')

    elif "list" in unstopped_tokens and "browsers" in unstopped_tokens or "list" in unstopped_tokens and "browser" in unstopped_tokens or "show" in unstopped_tokens and "browser" in unstopped_tokens:
        print("\033[1;34;1m")
        a = os.popen("pacaur -Qqs web browser").read().split("\n")[:20] # Search and index first 20 results
        for i, j in enumerate(a): # display the indexed file
            d[i] = j.split("\n")[0]
            print(d[i])

    elif "list" in unstopped_tokens and "image" in unstopped_tokens or "list" in unstopped_tokens and "image" in unstopped_tokens or "show" in unstopped_tokens and "image" in unstopped_tokens:
        print("\033[1;34;1m")
        a = os.popen("pacaur -Qqs image viewer").read().split("\n")[:20] # Search and index first 20 results
        for i, j in enumerate(a): # display the indexed file
            d[i] = j.split("\n")[0]
            print(d[i])

    elif "partitions" in unstopped_tokens and "system" in unstopped_tokens or "harddisk" in unstopped_tokens and "partitions" in unstopped_tokens or "hdd" in unstopped_tokens and "partitions" in unstopped_tokens:
        print("\033[1;34;1m")
        print("\n")
        os.system("lsblk -f")

    elif "internet" in tokens and "speed" in tokens:
        speedtest()

    elif "new" in unstopped_tokens and "mail" in tokens or "new" in unstopped_tokens and "email" in tokens or "new" in unstopped_tokens and "e-mail" in tokens:
        print("\033[1;34;1m")
        confirm = input("Do you wish to send a new email? ")
        if confirm == 'y' or confirm == 'yes' or confirm == 'ya' or confirm == 'yeah' or confirm == 'yep':
            mail()

    elif "kernel" in unstopped_tokens and "version" in unstopped_tokens or "which" in unstopped_tokens and "kernel" in unstopped_tokens:
        print("\033[1;34;1m")
        os.system("uname -r")  

    elif "my" in unstopped_tokens and "name" in unstopped_tokens:
        if callme == "there" or not callme:
            callme = boss()
        else:
            print("\033[1;34;1m")
            print("\nHazel : Come on, ",callme, "stop testing our relationship! I know who you are.")

    elif "memory" in unstopped_tokens and "usage" in tokens or"memory" in tokens and "available" in tokens or "storage" in tokens and "space" in tokens or "harddisk" in tokens and "space" in tokens and "left" in tokens:
        meminfo()

    elif "Which" in tokens and "os" in tokens and "i" in tokens or "Which" in tokens and "operating system" in tokens and "i" in tokens:
        platform.linux_distribution()

    elif "my" in unstopped_tokens and "os" in tokens or "os" in tokens and "using" in tokens or "operating system" in tokens and "using" in tokens:
        platform.linux_distribution()

    elif "existing" in tokens and "users" in tokens or "list" in tokens and "users" in tokens:
        list_users(tokens)

    elif "new" in tokens and "user" in tokens or "new" in tokens and "account" in tokens:
        adduser(tokens)

    elif "change" in tokens and "password" in tokens:
        chpasswd(tokens)

    elif "remove" in tokens and "user" in tokens or "remove" in tokens and "account" in tokens or "delete" in tokens and "user" in tokens or "delete" in tokens and "account" in tokens:
        deltuser(tokens)

    elif "send" in unstopped_tokens and "mail" in tokens or "send" in unstopped_tokens and "email" in tokens or "send" in unstopped_tokens and "e-mail" in tokens:
        print("\033[1;34;1m")
        confirm = input("Do you wish to send a new email? ")
        if confirm == 'y' or confirm == 'yes' or confirm == 'ya' or confirm == 'yeah' or confirm == 'yep':
            mail()

    elif "how" in unstopped_tokens and "to" in unstopped_tokens:
        socli(chat, unstopped_tokens)

    elif "how" in unstopped_tokens and "can" in unstopped_tokens and "i" in unstopped_tokens:
        socli(chat, unstopped_tokens)

    elif "how" in unstopped_tokens and "will" in unstopped_tokens and "i" in unstopped_tokens:
        socli(chat, unstopped_tokens)

    elif "my" in unstopped_tokens and "ip" in tokens or "my" in unstopped_tokens and "network" in tokens and "address" in tokens or "system" in tokens and "ip" in tokens:
        try:
            print("\033[1;34;1m")
            s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            s.connect(("8.8.8.8", 80))
            print("\nHazel : Your current IP address is", s.getsockname()[0])
            s.close()
        except Exception as e:
            print("\033[1;34;1m")
            print("no connection")
            s.close()

    elif "where" in unstopped_tokens and "is" in unstopped_tokens:
        wiki(unstopped_tokens, message)

    elif "get" in unstopped_tokens and "details" in unstopped_tokens:
        wiki(unstopped_tokens, message)

    elif "tell" in unstopped_tokens and "me" in unstopped_tokens and "about" in unstopped_tokens:
        wiki(unstopped_tokens, message)

    elif "get" in unstopped_tokens and "information" in unstopped_tokens:
        wiki(unstopped_tokens, message)
 
    elif "what" in unstopped_tokens and "is" in unstopped_tokens and "my" not in unstopped_tokens:
        socli(chat, unstopped_tokens)

    elif "call" in unstopped_tokens and "me" in unstopped_tokens:
        if callme == 'there' or not callme:
            callme = boss()
        else:
            print(callme)

    elif "internet" in tokens and "active" in tokens or "internet" in tokens and "connected" in tokens or "available" in tokens:
        is_connected()

    elif "who" in unstopped_tokens and "is" in unstopped_tokens or "who" in unstopped_tokens and "was" in unstopped_tokens:
        wiki(unstopped_tokens, message)


    ### GOOGLE SEARCH ###

    elif "google" in tokens: # googler api search to display top 10 results from google with link, links open in default browser 
        cmd = "googler"
        print("To quit the search results, type q")
        os.system(cmd + " " + message) # perform command in shell


    ### Close the application, ctrl+c or ctrl+z wont work ###

    elif "close" in tokens or "q" in tokens or 'quit' in tokens or 'exit' in tokens:
        exit(0)

    ### OPEN FILE MANAGER ###
    ### CURRENTLY SUPPORTS OPENING TEXT FILES OR PROGRAMMING FILES ###

    elif "create" in tokens:
        openfiles()



    #### IF USER INTENDS TO DO NONE OF THE ABOVE ACTIONS, RUN THE hazel_chatter.py FILE TO GET NORMAL CHAT OUTPUTS ###
    ### DEEP LEARNING MODEL NOT IMPLEMENTED YET. CURRENTLY WORKS WITH A TEXT DATABASE TO RETRIEVE AN NLP OUTPUT, THIS IS RECURSIVE CHAT AND NOT GENERATIVE AS IN DEEP LEARNING ###

    else:
        chatter(chat) # Function call to hazel_chatter.py, passed the original user input as asrguement



########################################################################################################################


### MAIN FUNCTION WHICH WORKS AT APPLICATION STARTUP AND RECEIVES USER INPUT ###

def main():

    print("\033[1;34;1m")
    print("\n\n\nHazel : Hey ",callme,", type 'h' to get help or simply start chatting with me") # Initial message shown at every startup
    
    SQLCompleter = WordCompleter(['install', 'uninstall', 'update', 'upgrade', 'who is', 'what is', 'quit', 'google', 'check internet connection', 'my ip address', 'my name', 'send email', 'internet speed', 'date today', 'time now', 'change date', 'change time', 'battery status'],
                             ignore_case=True)

    while True: # Enter the loop anyways
        print("\033[1;32;1m")
        message =   prompt('\nYou   : ',
                        history=FileHistory('/tmp/history.txt'),
                        auto_suggest=AutoSuggestFromHistory(),
                        completer=SQLCompleter,
                        )
        message = message.lower()
        msg = message.lower() # temporary storage of message value
        unstopped_tokens = ""
        tokens = word_tokenize(message) # NLTK function tokenize to tokenize user input
        unstopped_tokens = word_tokenize(message)
        d = {} # Store package names during package management
        tok = "" # Temporary variable to store tokens and to determine its value
        stop_words = set(stopwords.words('english')) # Remove stop words
        filtered_sentence = [w for w in tokens if not w in stop_words]
        filtered_sentence = []
        for w in tokens: # Filtering input by removing stopwords such as 'I', 'for', 'is', etc.
            if w not in stop_words:
                filtered_sentence.append(w) # Get and store message without stopeords

        filtered_list = filtered_sentence
        filtered_sentence = ' '.join(filtered_list) # Making a sentance out of the tokens
        chat = message
        message = filtered_sentence # storing input in message
        tokens = word_tokenize(message) # tokenize new message
        pcount = 0 # pcount to check if package manager search result is null or greater than 1

        message_length = len(message.split()) # Textblob function to check the sentiment asscoiated with a sentance.
        blob = TextBlob(message) # Apply sentiment analysis on message. If polarity is +ve or -ve, directly goto chat else if its neutral, perform system command or else do chat
        length = len(message.split()) # find length of the message

        for sentence in blob.sentences: # Get polarity or sentment value. 0 = neutral, -ve = negative, +ve = positive
            pol = sentence.sentiment.polarity # polarity ranges from -1 to +1

        if "h" in tokens: # Print the help.txt file to read the description and manual page of Hazel
            print("\033[1;37;1m")
            os.system("most /opt/hazel-linux_chatbot_and_assistant/hazel/help.txt") # Using 'most' command to print in pages 

        elif "h" not in tokens: # if not h, go through package list anyways to ckeck if user have mentioned the name of a package in the chat input

            tokened = tokens # Temporary variable for tokens


            ### STARTING FROM HERE IS PACKAGE MANAGEMENT, THIS IS UNDERGONE FOR ALL INPUTS TO MAKE SURE PACKAGE MANAGEMENT IS DONE IN THE MOST EFFICIENT WAY POSSIBLE, IF NO PACKAGE NAME OR ACTION IS MENTIONED, THE EXECUTION CONTINUES TO OTHER FUNCTIONS OR CHAT ###

            for i in range(0, len(tokened)): # search all tokened value
                if "install" in tokened:
                    tok = "install" # tok is set to install so that, all keywords that mey seem similar to install are passed to installer function
                    tokened.remove("install") # remove the checked keyword

                elif "dont" in tokened:
                    if "want" in tokened:
                        i = i + 1 # update i value as 1st value is checked and removed
                        tok = "remove" # tok is set to remove so that, all keywords that may seem similar to remove are passed to remover function
                        tokened.remove("dont") # remove the checked keyword
                        tokened.remove("want")
                        break
                    tok = "remove"
                    tokened.remove("want")

                elif "want" in tokened and "install" in tokened:
                    tok = "install"
                    tokened.remove("want")

                elif "want" in tokened and "remove" in tokened or "want" in tokened and "remove" in tokened:
                    tok = "remove"
                    tokened.remove("remove")

                elif "remove" in tokened:
                    tok = "remove"
                    tokened.remove("remove")


                elif "uninstall" in tokened:
                    tok = "remove"
                    tokened.remove("uninstall")

                elif "upgrade" in tokened:
                    print(tok)
                    if "system" in tokened:
                        i = i + 1
                        tok = "upgrade"
                        tokened.remove("system")
                    if "all" in tokened:
                        i = i + 1
                        tok = "upgrade"
                        tokened.remove("all")
                    if "apps" in tokened:
                        i = i + 1
                        tok = "upgrade"
                        tokened.remove("apps")
                    tokened.remove("upgrade")
                    update(message) # call update function in package_surfer.py

                elif "update" in tokened:
                    print(tok)
                    if "system" in tokened:
                        i = i + 1
                        tok = "upgrade"
                        tokened.remove("system")
                    if "all" in tokened:
                        i = i + 1
                        tok = "upgrade"
                        tokened.remove("all")
                    if "apps" in tokened:
                        i = i + 1
                        tok = "upgrade"
                        tokened.remove("apps")
                    tokened.remove("update")
                    update(message) # call update function in package_surfer.py


                elif "longer" in tokened:
                    if "need" in tokened:
                        i = i + 1
                        tok = "remove"
                        tokened.remove("need")
                    if "want" in tokened:
                        i = i + 1
                        tok = "remove"
                        tokened.remove("want")
                    tok = "remove"
                    tokened.remove("longer")
                    

            ###  END OF CONDITION CHECKING ###
            ### ENTERING PACKAGE NAME CROSS CHECKING WITH THE ONES IN THE DISTRIBUTION REPOSITORIES ###

            for j in range(0, len(tokened)):
                with open('/opt/hazel-linux_chatbot_and_assistant/hazel/repo_app_list.txt','r') as f: # Open repo_app_list.txt which contains the list of all possible packages that can be installed on this Linux
                    for line in f:
                        liner = line.strip() # remove white spaces
                        matcher = SequenceMatcher(None, liner, tokened[j]).ratio() # Match the list element with input term to see if a package has atleast 50% similar name as with the package in the repository

                        if matcher >= 0.75: # if a package with >= 50% match is found, it is listed later
                            pcount = pcount + 1 # updates pcount since a package with match is found

            if pcount > 0: # if atleast 1 package exists in repo_app_list.txt
                print("\033[1;34;1m")

                if "install" == tok: # tok is install
                    print("\nPackage Installer\n")
                    install(message) # call install function in package_surfer.py

                elif "remove" == tok: # tok is remove
                    print("\nPackage Remover\n")
                    remove(message) # call remove function in package_surfer.py

                else: # If none of the above operation is triggered, it is not package management, redirect to other funtions
                    others(tokens, message, unstopped_tokens, chat)
            else:
                others(tokens, message, unstopped_tokens, chat)

if __name__ == "__main__":
    main()
else:
    print("hell")

#main() # start executing the main.py file by calling main() funtion



##########################################################################################################################
